### PR TITLE
revert #9799 Deprecate scope as a type constraint on class declarations

### PR DESCRIPTION
**Pull request by WalterBright**
Reverts https://github.com/dlang/dmd/pull/9799 which is causing expensive problems for one of our users.

### CODE REVIEW

**Comment by Geod24** [comment]
Can you expand on the problems this is causing ?

**Comment by WalterBright** [comment]
The message: "Deprecation: scope as a type constraint is deprecated.  Use scope at the usage site." implies the user has control over the code at the use site. Worse, the code will still work (but perform poorly) if the user neglects to add it at the use site.

**Comment by WalterBright** [comment]
Well, this is bizarre. circleci says that test/fail_compilation/scope_test.d has a problem, but there is no scope_test.d in my repo.

**Comment by thewilsonator** [comment]
`fail_compilation/scope_type.d` not `scope_test.d`

**Comment by WalterBright** [comment]
That file isn't there, either :-/

**Comment by thewilsonator** [comment]
Look harder: https://github.com/dlang/dmd/blob/master/test/fail_compilation/scope_type.d

**Comment by WalterBright** [comment]
It's not in my revert9799 repo. How is that possible?

**Comment by WalterBright** [comment]
```
mercury~/forks/dmd/test/fail_compilation revert9799> ls scope*
scope_class.d
mercury~/forks/dmd/test/fail_compilation revert9799>
```

**Comment by thewilsonator** [comment]
Probably because the parent commit of this PR is ~8 months old an that was added after.

**Comment by thewilsonator** [comment]
Try rebasing this on the latest commit of whichever branch you are targeting.

**Comment by WalterBright** [comment]
What I want to do is convert my local revert9799 to match the stable revert9799. How do I do that? My various attempts have all failed.

**Comment by thewilsonator** [comment]
If you can't get it to work, it's a 9 line diff. I'd just do it manually.

**Comment by WalterBright** [comment]
Works for me. Thank you.

**Comment by WalterBright** [comment]
How do I copy this PR into a branch on my repo?

**Comment by thewilsonator** [comment]
```
git fetch origin #or whatever you `git remote -v` is for the your GH repo
git checkout revert9799
```

**Comment by Geod24** [comment]
> The message: "Deprecation: scope as a type constraint is deprecated. Use scope at the usage site." implies the user has control over the code at the use site. Worse, the code will still work (but perform poorly) if the user neglects to add it at the use site.

I mean the entire point of the deprecation was that the class definition should not be responsible for the allocation strategy. The same rationale applied to class (de)allocators.

As for the assumption that the user has control over the code at the use site, well, there is *no change* needed at the use site, because `scope` is already used (otherwise the code wouldn't compile).
So it's just a matter of removing `scope` on the declaration (and nothing change *for code not under active development*).

Note that the idea to deprecate it [came from you](https://github.com/dlang/dlang.org/pull/2027#issuecomment-374493221) after heavy discussion.

**Comment by WalterBright** [comment]
> Note that the idea to deprecate it came from you after heavy discussion.

It's incredibly rare, but sometimes I can be wrong.

**Comment by WalterBright** [comment]
ok, I tried:
```
git fetch origin
git checkout revert9799
```
and it says I changed every single file in the new branch.

Please explain EXACTLY what command I need to use. All I want to do is get THIS branch into MY repo.

**Comment by thewilsonator** [comment]
Right that should have had a `git rebase upstream/stable` in between. So

```
git fetch upstream
git checkout stable
git rebase upstream/stable
git checkout origin/revert9799
git checkout -b revert9799
```

**Comment by WalterBright** [comment]
Ok, this is looking good. Thanks, @thewilsonator

**Comment by don-clugston-sociomantic** [comment]
> I mean the entire point of the deprecation was that the class definition should not be responsible for the allocation strategy. The same rationale applied to class (de)allocators.

actually what this change means is:
`the class definition should not be able to impose restrictions on the allocation strategy.`

`scope` has a pretty similar meaning to `@nogc`. Its purpose is to detect bugs.
And in this case the bugs it detects are crucial to us.
It'd arguably be more appropriate if it was named `@scope`. Throwing that out a suggestion -- would that achieve any of the goals of removing 'scope class' ?

Now maybe in the future it would be possible to eliminate the need for it, if we had @nogc everywhere. Unfortunately we're not there yet.

**Comment by don-clugston-sociomantic** [comment]
LGTM.
Thanks, Walter.
I will try to increase the fraction of our code which uses nogc, so that this PR can be reinstated in the future. I suspect that the feature is rarely used outside of our codebase.

**Comment by don-clugston-sociomantic** [comment]
I didn't do this correctly. I merged instead of using the 'auto-merge' tag. Can the 'merge' button be disabled to prevent this?

**Comment by WalterBright** [comment]
The message merged it into dlang:stable. I think this was correct. What went wrong?

**Comment by don-clugston-sociomantic** [comment]
@Geod24 This wasn't merged properly, I think. It is not present in the nightly build of 2.091. This unfortunately means there is no branch with both of the patches that I need...

**Comment by Geod24** [comment]
Stable was just merged into master today. It'll be the nightly tomorrow.

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -1226,7 +1226,8 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
             // @@@DEPRECATED@@@  https://dlang.org/deprecate.html#scope%20as%20a%20type%20constraint
             // Deprecated in 2.087
             // Remove this when the feature is removed from the language
-            if (!(dsym.storage_class & STC.scope_))
+            if (0 &&          // deprecation disabled for now to accommodate existing extensive use
+               !(dsym.storage_class & STC.scope_))
```
Comment (ibuclaw) [inline src/dmd/dsymbolsem.d:1230]: @WalterBright why did you make this change?  This wasn't in the original PR in #9799.

**Comment by ibuclaw** [comment]
This PR caused a regression https://issues.dlang.org/show_bug.cgi?id=22780

**Pull request merged**